src/gui.rs

```diff
@@ -394,195 +375,26 @@ impl Sam {
-                    if let Some(data_source) = open_log_file() {
-                        self.open_log_file(data_source);
-                    }
-                }
-
-                // Toggle archive panel
-                ui.toggle_value(&mut self.archive_window_open, "🗄 Flight Archive");
-
-                // Toggle archive panel
-                if ui.selectable_label(self.data_source.simulation_settings().is_some(), "💻 Simulate").clicked() {
-                    self.data_source = Box::new(SimulationDataSource::default());
-                }
-
-                // Show a button to the right to close the current log/simulation and go back to
-                // live view
-                ui.allocate_ui_with_layout(ui.available_size(), Layout::right_to_left(Align::Center), |ui| {
-                    if self.data_source.is_log_file() || self.data_source.simulation_settings().is_some() {
-                        if ui.button("❌").clicked() {
-                            self.close_data_source();
-                        }
-                    }
-                });
-            });
-        });
+        create_top_menu_bar(self, ctx);
 
         // A window to open archived logs directly in the application
         let mut archive_open = self.archive_window_open; // necessary to avoid mutably borrowing self
-        egui::Window::new("Flight Archive").open(&mut archive_open).min_width(300.0).anchor(Align2::CENTER_CENTER, [0.0, 0.0]).resizable(false).collapsible(false).show(ctx, |ui| {
-            ui.add_space(10.0);
-
-            for (i, (title, telem, flash)) in ARCHIVE.iter().enumerate() {
-                if i != 0 {
-                    ui.separator();
-                }
-
-                ui.horizontal(|ui| {
-                    ui.label(*title);
-                    ui.with_layout(Layout::right_to_left(Align::Center), |ui| {
-                        if ui.add_enabled(flash.is_some(), Button::new("🖴  Flash")).clicked() {
-                            self.open_archive_log(flash.unwrap());
-                        }
-
-                        if ui.add_enabled(telem.is_some(), Button::new("📡 Telemetry")).clicked() {
-                            self.open_archive_log(telem.unwrap());
-                        }
-                    });
-                });
-            }
-
-            ui.add_space(10.0);
-            ui.horizontal(|ui| {
-                ui.add_visible_ui(self.archive_progress.is_some(), |ui| {
-                    let (done, total) = self.archive_progress.unwrap_or((0, 0));
-                    let f = (total > 0).then(|| done as f32 / total as f32).unwrap_or(0.0);
-                    let text = format!("{:.2}MiB / {:.2}MiB", done as f32 / (1024.0*1024.0), total as f32 / (1024.0*1024.0));
-                    ui.add_sized([ui.available_width(), 20.0], ProgressBar::new(f).text(text));
-                });
-            });
-            ui.add_space(10.0);
+        open_archive_window(ctx, &mut archive_open, self);
```

Comment:
Similar to above, great fan of moving some of the more specific UI stuff out of here though. For windows it might be interesting to have a system where windows can return values, in this case a flight log, to `src/gui.rs`. Something along the lines of:

```rust
if let Some(log) = ui.add(archive_window).result() {
    self.open_archive_log(log);
}
```

Same as above though, this isn't necessarily urgent.